### ISSUE TITLE
Clang thread-safety analysis not producing warnings for multiple objects of the same type

### ISSUE THREAD

**Issue by Qwinci** [llvm/llvm-project#127349]
Given the following code (full code in godbolt link below) Clang doesn't report any warnings for accessing b.some_field even though b.lock is not locked at any point (removing a.lock.acquire() + a.lock.release() calls makes the warning appear but that's obviously not something that you should have to do as they are two distinct objects):
```cpp
struct CAPABILITY("spinlock") Spinlock {
    void acquire() ACQUIRE();
    void release() RELEASE();
};

struct Foo {
    Spinlock lock;
    bool some_field GUARDED_BY(lock);
};

void foobar() {
    Foo a {};
    Foo b {};
    a.lock.acquire();
    b.some_field = true;
    a.lock.release();
}
```
[godbolt](https://godbolt.org/z/dPqb6aqrx)

**Comment by gle8098** [comment]
That's because by default the analysis checks capabilities in a relaxed way to reduce false positives. In your case `a.lock` partially matches `b.lock` since they share the same declaration so it is considered to be probably the same lock, although it is clearly a false negative here.

To disable partial matching add the `-Wthread-safety-precise` flag. [That would work for you](https://godbolt.org/z/frejso7aT).
